PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,449 @@
 Code related to processing of import hooks.
 """
 
-import glob
+import glob, weakref
 import os.path
 
 from .. import log as logging
-from .utils import format_binaries_and_datas
-from ..compat import expand_path
-from ..compat import importlib_load_source
+from ..compat import expand_path, importlib_load_source, FileNotFoundError
 from .imphookapi import PostGraphAPI
+from .utils import format_binaries_and_datas
 
 logger = logging.getLogger(__name__)
 
 
+class ModuleHookCache(dict):
```

Comment:
> The real winners are probably list catenation, set intersection, difference, dict update and the like.

Right. We definitely want all of those to be optimized to the hilt. `</sigh>`

> ...it will call the C-level functions like `PyDict_Next` when possible, instead of going through the Python method lookup and getting `iter(obj)`...

It's a rational change from the limited perspective of optimization gains. But this is CPython, in which multithreading – the most primitive building block of parallelization, available in most modern languages out-of-the-box – is infeasible. <sup>...for all intents and purposes.</sup>

**Urgh.** Ruby definitely got it right on this one. Ruby also has a variety of working JIT interpreters coming online (e.g., B3, Rubinius). _Oh, well._